Add friends likes

SergeyLeshchev opened the pull request.

SergeyLeshchev committed:
Выполнить ТЗ №11

SergeyLeshchev committed:
Выполнить ТЗ №11

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -2,88 +2,55 @@
+    public FilmController(FilmStorage filmStorage, FilmService filmService, UserStorage userStorage) {
+        this.filmStorage = filmStorage;
+        this.filmService = filmService;
+    }
```
Вместо явного объявления контроллеров можно использовать аннотацию `@RequiredArgsConstructor`, если так больше нравится)

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -2,88 +2,55 @@
-    private final Map<Long, Film> films = new HashMap<>();
-    private long idCounter = 1;
+    private final FilmStorage filmStorage;
+    private final FilmService filmService;
```
У нас используется архитектура, по которой взаимодействие выстраивается по цепочке controller -> service -> repository, соответственно в контроллерах мы обращаемся к сервисам, в сервисах - к репозиториям. Это задаст жесткую зону ответственности для наших слоев. Контроллер занимается обработкой входящего запроса и передает его в сервис, сервис содержит и выполняет основную логику, а репозиторий работает с сущностями и их хранением.

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -1,38 +1,56 @@
-    // 28 декабря 1895 года считается днём рождения кино
+    /*
+     *
+     * Я закомментировал аннотации над полями так как с ними у меня не проходили тесты в Postman, а без аннотаций проходят.
+     * Так как с аннотациями у меня не появляется ни одной строки написанного мной лога о том, что программа вошла в метод,
+     * то я думаю, что различные исключения выбрасываются во время проверки соответствия тела запроса и конструктора объекта.
+     * Например, тест на проверку логина не проходит, когда нет имени, статус 500. Если имя есть, то программа заходит в
+     * метод валидации и выбрасывает нужное исключение и статус 400.
+     *
+     * Похожая ситуация и с датой рождения. Есть тест с датой рождения пользователя в будущем. Этот пользователь не
+     * создается и программа падает еще до начала работы логики метода.
```
Валидация отрабатывает еще до начала работы метода контроллера, спринг при ошибках валидации оборачивает их в `MethodArgumentNotValidException`, если напишешь для него обработчик в `ErrorHandler`, сможешь возвращать нужный статус и сообщение, тогда с тестами проблема будет решена

SergeyLeshchev replied:
О, да, это то что надо

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -1,38 +1,56 @@
+        this.releaseDate = releaseDate;
+        this.duration = duration;
+        this.rate = rate;
+        // По какой-то причине во время создания объекта это поле остается null. Думаю это потому что @RequestBody
+        // использует конструктор без параметров. Я решил вынести инициализацию в начало класса вместо переопределения
+        // конструктора, думаю так надежнее. Не знаю как лучше
+        // likes = new HashSet<>();
```
Все верно, можно реализовать и конструктор без параметров (`@NoArgsConstructor`) и конструктор со всеми параметрами  (`@AllArgsConstructor`)  если необходимо

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,57 @@
+    public String addLike(Long filmId, Long userId) {
+        log.info("addLike() - получен запрос от пользователя с id {} на добавление лайка фильму с id {}",
+                userId, filmId);
+        if (!filmStorage.getFilmsMap().containsKey(filmId) || !userStorage.getUsersMap().containsKey(userId)) {
+            throw new NotFoundException("Не найден пользователь с таким id");
+        }
+        filmStorage.getFilm(filmId).getLikes().add(userId);
```
Будь это не хешмапы, а БД, ты бы тут обращался к ней вместо одного раза несколько для различных проверок и получения сущности. Это не есть гуд и нужно всегда минимизировать нагрузку БД.

Проще всего реализовать проверку в методе `getFilm()` и возвращать объект если он существует, иначе выбрасывать исключение. Тогда все эти лишние строки с проверками будут не нужны.

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,57 @@
+        filmStorage.getFilm(filmId).getLikes().add(userId);
+        log.info("addLike() - выполнен запрос от пользователя с id {} на добавление лайка фильму с id {}",
+                userId, filmId);
+        return ""; //"Вы поставили лайк фильму " + filmStorage.getFilm(filmId).getName();
```
Зачем вообще, тогда что-то возвращать? Пусть будут войдовыми

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
Тут аналогичные моменты с проверками и возвратами пустых строк

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/storage/InMemoryFilmStorage.java:
```diff
@@ -0,0 +1,55 @@
+        return films.get(id);
+    }
+
+    public Map<Long, Film> getFilmsMap() {
+        return films;
+    }
```
Методы возвращающие хешмапу фильмов/юзеров не нужны, всегда можно получить конкретную сущность по айди (getFilm()) или все (getFilms())

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/validator/FilmValidator.java:
```diff
@@ -0,0 +1,46 @@
+        }
+
+        // максимальная длина описания — 200 символов
+        if (film.getDescription().length() > 200) {
+            log.warn("Исключение ValidationException. Максимальная длина описания — 200 символов");
+            throw new ValidationException("Максимальная длина описания — 200 символов");
+        }
+
+        // дата релиза — не раньше 28 декабря 1895 года
+        // 28 декабря 1895 года считается днём рождения кино
+        final LocalDate BIRTHDAY_FILM = LocalDate.parse("1895-12-28");
+        if (film.getReleaseDate().isBefore(BIRTHDAY_FILM)) {
+            log.warn("Исключение ValidationException. Дата релиза — должна быть не раньше 28 декабря 1895 года");
+            throw new ValidationException("Дата релиза — должна быть не раньше 28 декабря 1895 года");
+        }
```
Если придут нуллы по этим полям, то мы упадем с NPE

SergeyLeshchev committed:
Выполнить работу над ошибками к ТЗ №11

1. Выстроить архитектуру controller -> service -> repository
2. Вернуть аннотации валидации полей в классах Film и User
3. Добавить обработку исключения MethodArgumentNotValidException
4. Доработать проверку на null в методах getFilm() и getUser()
5. Исправить методы, которые возвращали пустую строку
6. Удалить методы, которые возвращали коллекции с сущностями
7. Добавить в методы валидации проверку на null

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/storage/InMemoryFilmStorage.java:
```diff
@@ -46,10 +46,13 @@ public Collection<Film> getFilms() {
-
-    public Map<Long, Film> getFilmsMap() {
-        return films;
+        if (!films.containsKey(id)) {
+            throw new NotFoundException("Нет фильма с таким id");
+        }
```
А зачем эта проверка? Достаточно того, что ниже)

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/storage/InMemoryUserStorage.java:
```diff
@@ -46,10 +46,13 @@ public Collection<User> getUsers() {
-
-    public Map<Long, User> getUsersMap() {
-        return users;
+        if (!users.containsKey(id)) {
+            throw new NotFoundException("Нет пользователя с таким id");
+        }
```
Тут тоже

SergeyLeshchev committed:
Выполнить работу над ошибками №2 к ТЗ №11

1. Убрать лишнюю проверку в методах getFilm и getUser
